mime.go

```diff
@@ -116,107 +123,203 @@ func (mh *mimeHandler) initJPEGEncoder(frame *C.AVFrame) error {
 	}
-	if mh.yuyvFrame == nil {
-		return nil, camera.ImageMetadata{}, errors.New("failed to create yuyv destination frame")
+	if *dstFrame == nil {
+		return nil, camera.ImageMetadata{}, fmt.Errorf("failed to create %s destination frame", format)
 	}
+
 	res := C.sws_scale(
-		mh.yuyvSwsCtx,
+		*swsCtx,
 		(**C.uint8_t)(unsafe.Pointer(&frame.data[0])),
 		(*C.int)(unsafe.Pointer(&frame.linesize[0])),
 		0,
 		frame.height,
-		(**C.uint8_t)(unsafe.Pointer(&mh.yuyvFrame.data[0])),
-		(*C.int)(unsafe.Pointer(&mh.yuyvFrame.linesize[0])),
+		(**C.uint8_t)(unsafe.Pointer(&(*dstFrame).data[0])),
+		(*C.int)(unsafe.Pointer(&(*dstFrame).linesize[0])),
 	)
 	if res < 0 {
-		return nil, camera.ImageMetadata{}, newAvError(res, "failed to convert frame to YUYV")
-	}
-	yuyvDataSize := int(mh.yuyvFrame.width) * int(mh.yuyvFrame.height) * yuyvBytesPerPixel
-	header := packYUYVHeader(int(mh.yuyvFrame.width), int(mh.yuyvFrame.height))
-	// Preallocate the final slice with the combined size of the header and the image data
-	yuyvPacket := make([]byte, len(header)+yuyvDataSize)
-	copy(yuyvPacket[0:], header)
-	// Copy the YUYV data directly into the preallocated slice
-	C.memcpy(unsafe.Pointer(&yuyvPacket[len(header)]), unsafe.Pointer(mh.yuyvFrame.data[0]), C.size_t(yuyvDataSize))
-
-	return yuyvPacket, camera.ImageMetadata{
-		MimeType: mimeTypeYUYV,
+		return nil, camera.ImageMetadata{}, newAvError(res, "failed to convert frame to "+format)
+	}
+
+	dataSize := int((*dstFrame).width) * int((*dstFrame).height) * bytesPerPixel
+	header := packHeader(format, int((*dstFrame).width), int((*dstFrame).height))
+	data := make([]byte, len(header)+dataSize)
+	copy(data[0:], header)
+	C.memcpy(unsafe.Pointer(&data[len(header)]), unsafe.Pointer((*dstFrame).data[0]), C.size_t(dataSize))
+
+	return data, camera.ImageMetadata{
+		MimeType: mimeType,
 	}, nil
 }
 
 func (mh *mimeHandler) initYUYVContext(frame *C.AVFrame) error {
-	mh.logger.Info("creating YUYV sws context with frame size: ", frame.width, "x", frame.height)
-	if mh.yuyvSwsCtx != nil {
-		C.sws_freeContext(mh.yuyvSwsCtx)
+	return mh.initPixelFormatContext(frame, C.AV_PIX_FMT_YUYV422, &mh.yuyvSwsCtx, &mh.yuyvFrame)
+}
+
+func (mh *mimeHandler) initRGBAContext(frame *C.AVFrame) error {
+	return mh.initPixelFormatContext(frame, C.AV_PIX_FMT_RGBA, &mh.rgbaSwsCtx, &mh.rgbaFrame)
+}
+
+// initPixelFormatContext is a helper function that initializes the conversion context and destination frame
+// for pixel format conversions. It handles cleanup of existing contexts/frames and allocation of new ones.
+//
+// Parameters:
+// - frame: Source AVFrame containing the input image
+// - pixFmt: Target pixel format to convert to
+// - swsCtxPtr: Pointer to SwsContext pointer that will be initialized
+// - dstFrame: Pointer to AVFrame pointer that will be initialized
+//
+// Returns error if any allocation or initialization fails
+func (mh *mimeHandler) initPixelFormatContext(
```

Comment:
Where is this called?

Reply:
Above by the yuyv/rgba specific context helper funcs